File: android/app/src/main/java/dev/talk2scale/MockScaleTransport.java
package dev.talk2scale;

import android.bluetooth.BluetoothProfile;

import java.util.Random;

public class MockScaleTransport implements ScaleTransport {

    private final Random random = new Random();
    private Listener listener;
    private int currentWeight;

    public void start() {
        dispatchConnectionState(BluetoothProfile.STATE_CONNECTED);
        dispatchWeightData(currentWeight, 0x01);
    }

    public void addRandomWeight() {
        currentWeight += random.nextInt(251) + 50;
        dispatchWeightData(currentWeight, 0x01);
    }

    @Override
    public void setListener(Listener listener) {
        this.listener = listener;
    }

    @Override
    public void sendTare() {
        currentWeight = 0;
        dispatchWeightData(currentWeight, 0x01);
    }

    @Override
    public void sendCalibrate(int refMassGrams) {
        // No-op for mock implementation.
    }

    @Override
    public void close() {
        dispatchConnectionState(BluetoothProfile.STATE_DISCONNECTED);
    }

    private void dispatchConnectionState(int state) {
        if (listener != null) {
            listener.onConnectionStateChanged(state);
        }
    }

    private void dispatchWeightData(int weight, int flags) {
        if (listener != null) {
            listener.onWeightData(weight, flags);
        }
    }
}

File: mobile/src/transport/MockScaleTransport.ts
import { ConnectionState, type ConnectionStateValue } from '../constants/ble';
import type { ScaleTransport, ScaleTransportListener } from './ScaleTransport';

export class MockScaleTransport implements ScaleTransport {
  private listener: ScaleTransportListener | null = null;
  private currentWeight = 0;

  setListener(listener: ScaleTransportListener | null): void {
    this.listener = listener;
  }

  start(): void {
    this.dispatchConnectionState(ConnectionState.CONNECTED);
    this.dispatchWeightData(this.currentWeight);
  }

  addRandomWeight(): void {
    this.currentWeight += Math.floor(Math.random() * 251) + 50;
    this.dispatchWeightData(this.currentWeight);
  }

  sendTare(): void {
    this.currentWeight = 0;
    this.dispatchWeightData(this.currentWeight);
  }

  sendCalibrate(_refMassGrams: number): void {
    // No-op for mock.
  }

  close(): void {
    this.dispatchConnectionState(ConnectionState.DISCONNECTED);
  }

  private dispatchConnectionState(state: ConnectionStateValue): void {
    this.listener?.onConnectionStateChanged(state);
  }

  private dispatchWeightData(weight: number): void {
    this.listener?.onWeightData(weight);
  }
}
